File: MatskocaAdmin/MatskocaAdmin/Components/StaffScheduleComponents.swift
import SwiftUI

struct StaffScheduleTabContent: View {
    let schedule: ScheduleTabUi
    let onDaySelected: (String) -> Void
    let onTypeFilterSelected: (String?) -> Void

    var body: some View {
        if schedule.denied {
            StaffEmptyState(message: schedule.deniedMessage)
                .padding(16)
        } else {
            VStack(spacing: 0) {
                if !schedule.days.isEmpty {
                    StaffScheduleDateSelector(days: schedule.days, onDaySelected: onDaySelected)
                }
                StaffScheduleTypeFilters(
                    selectedFilter: schedule.selectedTypeFilter,
                    onFilterSelected: onTypeFilterSelected
                )

                if schedule.loading {
                    Spacer()
                    ProgressView().tint(StaffColors.primary)
                    Spacer()
                } else if schedule.sessions.isEmpty {
                    VStack(spacing: 16) {
                        Image(systemName: "calendar.badge.exclamationmark")
                            .font(.system(size: 64))
                            .foregroundStyle(StaffColors.onSurfaceVariant)
                        Text("Нет тренировок на выбранную дату")
                            .font(.body)
                            .foregroundStyle(StaffColors.onSurfaceVariant)
                    }
                    .frame(maxWidth: .infinity, maxHeight: .infinity)
                    .padding(32)
                } else {
                    ScrollView {
                        LazyVStack(spacing: 12) {
                            ForEach(schedule.sessions) { session in
                                StaffScheduleSessionCard(session: session)
                            }
                        }
                        .padding(16)
                    }
                }
            }
        }
    }
}

private struct StaffScheduleDateSelector: View {
    let days: [ScheduleDayUi]
    let onDaySelected: (String) -> Void

    var body: some View {
        ScrollView(.horizontal, showsIndicators: false) {
            HStack(spacing: 8) {
                ForEach(days) { day in
                    StaffScheduleDateItem(day: day) {
                        onDaySelected(day.date)
                    }
                }
            }
            .padding(.horizontal, 8)
            .padding(.vertical, 12)
        }
        .background(StaffColors.surface)
    }
}

private struct StaffScheduleDateItem: View {
    let day: ScheduleDayUi
    let onClick: () -> Void

    var body: some View {
        Button(action: onClick) {
            VStack(spacing: 4) {
                Text(day.weekdayLabel)
                    .font(.caption2)
                    .opacity(0.8)
                Text(day.dayNumber)
                    .font(.title3)
                    .fontWeight(.bold)
                if day.isToday {
                    Circle()
                        .fill(day.selected ? StaffColors.onPrimary : StaffColors.primary)
                        .frame(width: 6, height: 6)
                }
            }
            .foregroundStyle(day.selected ? StaffColors.onPrimary : StaffColors.onSurfaceVariant)
            .padding(.horizontal, 16)
            .padding(.vertical, 12)
            .background(day.selected ? StaffColors.primary : Color(.systemGray5))
            .clipShape(RoundedRectangle(cornerRadius: 12))
        }
        .buttonStyle(.plain)
    }
}

private struct StaffScheduleTypeFilters: View {
    let selectedFilter: String?
    let onFilterSelected: (String?) -> Void

    var body: some View {
        ScrollView(.horizontal, showsIndicators: false) {
            HStack(spacing: 8) {
                filterChip("Все", selected: selectedFilter == nil) { onFilterSelected(nil) }
                filterChip("Групповые", selected: selectedFilter == "group") { onFilterSelected("group") }
                filterChip("Персональные", selected: selectedFilter == "personal") { onFilterSelected("personal") }
                filterChip("Допуслуги", selected: selectedFilter == "extra") { onFilterSelected("extra") }
            }
            .padding(.horizontal, 16)
            .padding(.vertical, 8)
        }
    }

    private func filterChip(_ label: String, selected: Bool, action: @escaping () -> Void) -> some View {
        Button(action: action) {
            Text(label)
                .font(.subheadline)
                .padding(.horizontal, 12)
                .padding(.vertical, 8)
                .background(selected ? StaffColors.primary : StaffColors.surface)
                .foregroundStyle(selected ? StaffColors.onPrimary : StaffColors.onBackground)
                .clipShape(Capsule())
                .overlay(Capsule().stroke(Color.gray.opacity(0.3), lineWidth: selected ? 0 : 1))
        }
        .buttonStyle(.plain)
    }
}

struct StaffScheduleSessionCard: View {
    let session: ScheduleSessionUi

    var body: some View {
        HStack(alignment: .top, spacing: 16) {
            VStack(spacing: 2) {
                Text(session.startTime)
                    .font(.title3)
                    .fontWeight(.bold)
                    .foregroundStyle(StaffColors.primary)
                Text("\(session.durationMinutes) мин")
                    .font(.caption)
                    .foregroundStyle(StaffColors.onSurfaceVariant)
            }

            RoundedRectangle(cornerRadius: 2)
                .fill(typeAccentColor(session.type))
                .frame(width: 3, height: 60)

            VStack(alignment: .leading, spacing: 4) {
                Text(session.typeLabel)
                    .font(.caption2)
                    .foregroundStyle(StaffColors.primary)
                Text(session.title)
                    .font(.headline)
                    .lineLimit(2)
                HStack(spacing: 4) {
                    Image(systemName: "person")
                        .font(.caption)
                    Text(session.trainer)
                        .font(.subheadline)
                        .lineLimit(1)
                }
                .foregroundStyle(StaffColors.onSurfaceVariant)
                HStack(spacing: 4) {
                    Image(systemName: "mappin.and.ellipse")
                        .font(.caption)
                    Text(session.room)
                        .font(.subheadline)
                }
                .foregroundStyle(StaffColors.onSurfaceVariant)
            }

            Spacer(minLength: 0)

            VStack(alignment: .trailing) {
                spotsLabelView
            }
        }
        .padding(16)
        .background(StaffColors.surface)
        .clipShape(RoundedRectangle(cornerRadius: 16))
        .shadow(color: .black.opacity(0.06), radius: 2, y: 1)
    }

    @ViewBuilder
    private var spotsLabelView: some View {
        if let label = sessionSpotsLabel(session) {
            if label.isFull {
                Text("Мест нет")
                    .font(.caption)
                    .fontWeight(.medium)
                    .padding(.horizontal, 8)
                    .padding(.vertical, 4)
                    .background(StaffColors.error.opacity(0.1))
                    .foregroundStyle(StaffColors.error)
                    .clipShape(Capsule())
            } else if let spotsLeft = label.spotsLeft {
                Text("Осталось \(spotsLeft)")
                    .font(.caption)
                    .fontWeight(.medium)
                    .foregroundStyle(spotsLeft <= 3 ? StaffColors.warning : StaffColors.onSurfaceVariant)
            } else if label.clientCount > 0 {
                Text("Записано \(label.clientCount)")
                    .font(.caption)
                    .fontWeight(.medium)
                    .padding(.horizontal, 8)
                    .padding(.vertical, 4)
                    .background(StaffColors.success.opacity(0.1))
                    .foregroundStyle(StaffColors.success)
                    .clipShape(Capsule())
            }
        }
    }

    private func typeAccentColor(_ type: String) -> Color {
        switch type {
        case "group": return StaffColors.accentBlue
        case "personal": return StaffColors.primary
        case "extra": return StaffColors.success
        default: return StaffColors.warning
        }
    }

    private struct SessionSpotsLabel {
        var spotsLeft: Int?
        var clientCount: Int = 0
        var isFull: Bool = false
    }

    private func sessionSpotsLabel(_ session: ScheduleSessionUi) -> SessionSpotsLabel? {
        if let booked = session.bookedCount, let maxParticipants = session.maxParticipants {
            let left = Swift.max(0, maxParticipants - booked)
            return SessionSpotsLabel(spotsLeft: left, isFull: booked >= maxParticipants)
        }
        if !session.clientNames.isEmpty {
            return SessionSpotsLabel(clientCount: session.clientNames.count)
        }
        return nil
    }
}

File: staffapp/src/main/java/com/example/staffapp/ui/work/StaffScheduleComponents.kt
package com.example.staffapp.ui.work

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.EventBusy
import androidx.compose.material.icons.filled.Person
import androidx.compose.material.icons.filled.Room
import androidx.compose.material3.AssistChip
import androidx.compose.material3.AssistChipDefaults
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.FilterChip
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import com.example.staffapp.ui.components.StaffEmptyState
import com.example.staffapp.ui.theme.StaffAccentBlue
import com.example.staffapp.ui.theme.StaffError
import com.example.staffapp.ui.theme.StaffPrimary
import com.example.staffapp.ui.theme.StaffSuccess
import com.example.staffapp.ui.theme.StaffWarning

@Composable
fun StaffScheduleTabContent(
    schedule: ScheduleTabUi,
    onDaySelected: (String) -> Unit,
    onTypeFilterSelected: (String?) -> Unit,
    onSessionClick: (ScheduleSessionUi) -> Unit = {},
) {
    if (schedule.denied) {
        Column(Modifier.fillMaxSize().padding(16.dp)) {
            StaffEmptyState(schedule.deniedMessage)
        }
        return
    }

    Column(modifier = Modifier.fillMaxSize()) {
        if (schedule.days.isNotEmpty()) {
            StaffScheduleDateSelector(
                days = schedule.days,
                onDaySelected = onDaySelected,
            )
        }
        StaffScheduleTypeFilters(
            selectedFilter = schedule.selectedTypeFilter,
            onFilterSelected = onTypeFilterSelected,
        )

        when {
            schedule.loading -> {
                Box(
                    modifier = Modifier.fillMaxSize(),
                    contentAlignment = Alignment.Center,
                ) {
                    CircularProgressIndicator(color = StaffPrimary)
                }
            }
            schedule.sessions.isEmpty() -> {
                Column(
                    modifier = Modifier
                        .fillMaxSize()
                        .padding(32.dp),
                    horizontalAlignment = Alignment.CenterHorizontally,
                    verticalArrangement = Arrangement.Center,
                ) {
                    Icon(
                        imageVector = Icons.Default.EventBusy,
                        contentDescription = null,
                        modifier = Modifier.size(64.dp),
                        tint = MaterialTheme.colorScheme.onSurfaceVariant,
                    )
                    Spacer(modifier = Modifier.height(16.dp))
                    Text(
                        text = "Нет тренировок на выбранную дату.\nНажмите +, чтобы создать запись.",
                        style = MaterialTheme.typography.bodyLarge,
                        color = MaterialTheme.colorScheme.onSurfaceVariant,
                    )
                }
            }
            else -> {
                LazyColumn(
                    contentPadding = PaddingValues(16.dp),
                    verticalArrangement = Arrangement.spacedBy(12.dp),
                ) {
                    items(schedule.sessions, key = { "${it.trainingId}_${it.startTime}_${it.title}" }) { session ->
                        StaffScheduleSessionCard(
                            session = session,
                            onClick = { onSessionClick(session) },
                        )
                    }
                }
            }
        }
    }
}

@Composable
private fun StaffScheduleDateSelector(
    days: List<ScheduleDayUi>,
    onDaySelected: (String) -> Unit,
) {
    val scrollState = rememberScrollState()
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .background(MaterialTheme.colorScheme.surface)
            .horizontalScroll(scrollState)
            .padding(vertical = 12.dp, horizontal = 8.dp),
        horizontalArrangement = Arrangement.spacedBy(8.dp),
    ) {
        days.forEach { day ->
            StaffScheduleDateItem(
                day = day,
                onClick = { onDaySelected(day.date) },
            )
        }
    }
}

@Composable
private fun StaffScheduleDateItem(
    day: ScheduleDayUi,
    onClick: () -> Unit,
) {
    val backgroundColor = if (day.selected) {
        MaterialTheme.colorScheme.primary
    } else {
        MaterialTheme.colorScheme.surfaceVariant
    }
    val contentColor = if (day.selected) {
        MaterialTheme.colorScheme.onPrimary
    } else {
        MaterialTheme.colorScheme.onSurfaceVariant
    }

    Column(
        modifier = Modifier
            .clip(RoundedCornerShape(12.dp))
            .background(backgroundColor)
            .clickable(onClick = onClick)
            .padding(horizontal = 16.dp, vertical = 12.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
    ) {
        Text(
            text = day.weekdayLabel,
            style = MaterialTheme.typography.labelSmall,
            color = contentColor.copy(alpha = 0.8f),
        )
        Text(
            text = day.dayNumber,
            style = MaterialTheme.typography.titleMedium,
            fontWeight = FontWeight.Bold,
            color = contentColor,
        )
        if (day.isToday) {
            Spacer(modifier = Modifier.height(4.dp))
            Box(
                modifier = Modifier
                    .size(6.dp)
                    .background(
                        color = if (day.selected) contentColor else MaterialTheme.colorScheme.primary,
                        shape = CircleShape,
                    ),
            )
        }
    }
}

@Composable
private fun StaffScheduleTypeFilters(
    selectedFilter: String?,
    onFilterSelected: (String?) -> Unit,
) {
    val chipScroll = rememberScrollState()
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .horizontalScroll(chipScroll)
            .padding(horizontal = 16.dp, vertical = 8.dp),
        horizontalArrangement = Arrangement.spacedBy(8.dp),
    ) {
        FilterChip(
            selected = selectedFilter == null,
            onClick = { onFilterSelected(null) },
            label = { Text("Все") },
        )
        FilterChip(
            selected = selectedFilter == "group",
            onClick = { onFilterSelected("group") },
            label = { Text("Групповые") },
        )
        FilterChip(
            selected = selectedFilter == "personal",
            onClick = { onFilterSelected("personal") },
            modifier = Modifier.widthIn(min = 110.dp),
            label = {
                Text(
                    "Персональные",
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                )
            },
        )
        FilterChip(
            selected = selectedFilter == "extra",
            onClick = { onFilterSelected("extra") },
            label = { Text("Допуслуги") },
        )
    }
}

@Composable
fun StaffScheduleSessionCard(
    session: ScheduleSessionUi,
    modifier: Modifier = Modifier,
    onClick: () -> Unit = {},
) {
    Card(
        modifier = modifier
            .fillMaxWidth()
            .clickable(onClick = onClick),
        shape = RoundedCornerShape(16.dp),
        colors = CardDefaults.cardColors(containerColor = MaterialTheme.colorScheme.surface),
        elevation = CardDefaults.cardElevation(defaultElevation = 2.dp),
    ) {
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(16.dp),
        ) {
            Column(horizontalAlignment = Alignment.CenterHorizontally) {
                Text(
                    text = session.startTime,
                    style = MaterialTheme.typography.titleMedium,
                    fontWeight = FontWeight.Bold,
                    color = MaterialTheme.colorScheme.primary,
                )
                Text(
                    text = "${session.durationMinutes} мин",
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }

            Spacer(modifier = Modifier.width(16.dp))

            Box(
                modifier = Modifier
                    .width(3.dp)
                    .height(60.dp)
                    .background(
                        color = typeAccentColor(session.type),
                        shape = RoundedCornerShape(2.dp),
                    ),
            )

            Spacer(modifier = Modifier.width(16.dp))

            Column(modifier = Modifier.weight(1f)) {
                Text(
                    text = session.typeLabel,
                    style = MaterialTheme.typography.labelSmall,
                    color = MaterialTheme.colorScheme.primary,
                )
                Text(
                    text = session.title,
                    style = MaterialTheme.typography.titleMedium,
                    fontWeight = FontWeight.SemiBold,
                    maxLines = 2,
                    overflow = TextOverflow.Ellipsis,
                )

                Spacer(modifier = Modifier.height(4.dp))

                Row(verticalAlignment = Alignment.CenterVertically) {
                    Icon(
                        imageVector = Icons.Default.Person,
                        contentDescription = null,
                        modifier = Modifier.size(16.dp),
                        tint = MaterialTheme.colorScheme.onSurfaceVariant,
                    )
                    Spacer(modifier = Modifier.width(4.dp))
                    Text(
                        text = session.trainer,
                        style = MaterialTheme.typography.bodyMedium,
                        color = MaterialTheme.colorScheme.onSurfaceVariant,
                        maxLines = 1,
                        overflow = TextOverflow.Ellipsis,
                    )
                }

                Spacer(modifier = Modifier.height(4.dp))

                Row(verticalAlignment = Alignment.CenterVertically) {
                    Icon(
                        imageVector = Icons.Default.Room,
                        contentDescription = null,
                        modifier = Modifier.size(16.dp),
                        tint = MaterialTheme.colorScheme.onSurfaceVariant,
                    )
                    Spacer(modifier = Modifier.width(4.dp))
                    Text(
                        text = session.room,
                        style = MaterialTheme.typography.bodyMedium,
                        color = MaterialTheme.colorScheme.onSurfaceVariant,
                    )
                }
            }

            Column(horizontalAlignment = Alignment.End) {
                sessionSpotsLabel(session)?.let { label ->
                    when {
                        label.isFull -> {
                            AssistChip(
                                onClick = {},
                                label = { Text("Мест нет") },
                                colors = AssistChipDefaults.assistChipColors(
                                    containerColor = StaffError.copy(alpha = 0.1f),
                                    labelColor = StaffError,
                                ),
                            )
                        }
                        label.spotsLeft != null && label.spotsLeft <= 3 -> {
                            Text(
                                text = "Осталось ${label.spotsLeft}",
                                style = MaterialTheme.typography.labelMedium,
                                color = StaffWarning,
                            )
                        }
                        label.spotsLeft != null -> {
                            Text(
                                text = "Осталось ${label.spotsLeft}",
                                style = MaterialTheme.typography.labelMedium,
                                color = MaterialTheme.colorScheme.onSurfaceVariant,
                            )
                        }
                        label.clientCount > 0 -> {
                            AssistChip(
                                onClick = {},
                                label = { Text("Записано ${label.clientCount}") },
                                colors = AssistChipDefaults.assistChipColors(
                                    containerColor = StaffSuccess.copy(alpha = 0.1f),
                                    labelColor = StaffSuccess,
                                ),
                            )
                        }
                    }
                }
            }
        }
    }
}

private data class SessionSpotsLabel(
    val spotsLeft: Int? = null,
    val clientCount: Int = 0,
    val isFull: Boolean = false,
)

private fun sessionSpotsLabel(session: ScheduleSessionUi): SessionSpotsLabel? {
    val booked = session.bookedCount
    val max = session.maxParticipants
    if (booked != null && max != null) {
        val left = (max - booked).coerceAtLeast(0)
        return SessionSpotsLabel(
            spotsLeft = left,
            isFull = booked >= max,
        )
    }
    if (session.clientNames.isNotEmpty()) {
        return SessionSpotsLabel(clientCount = session.clientNames.size)
    }
    return null
}

private fun typeAccentColor(type: String): Color = when (type) {
    "group" -> StaffAccentBlue
    "personal" -> StaffPrimary
    "extra" -> StaffSuccess
    else -> StaffWarning
}
